### PR TITLE
Zephyr standalone platform update

### PR DESCRIPTION
**Pull request by QuentinCaldeira-eaton**
As discussed in #33471 and #33777, the purpose of this PR is to leave open the possibility of using the Zephyr platform as a standalone platform, working with any boards present in Zephyr (which has network connectivity, enough flash, etc...). For testing purposes, I use a nucleo h563zi. This standalone platform **should** be used for manufacturers working on Zephyr, but also for people which would like to implement Matter in an unsupported board, using Zephyr OS. This zephyr "module" should simplify developer's life on the Zephyr platform.

Currently, zephyr platform can't be used in a Zephyr ecosystems without modifications, that's why I made that. My guess is anybody which has his board into Zephyr, like ESP or ST, could use the zephyr standalone platform to develop a product. With this, it is not mandatory to use a specific manufacturer zephyr sdk.

It would be great to also develop an example (like lighting-app) which work for the Zephyr SDK, but my knowledge on the Matter stack are very limited currently.

To sum-up, this PR make changes in order to have a functional connectedhomeip, as a Zephyr sdk module.

### CODE REVIEW

File: src/platform/Zephyr/BLEManagerImpl.cpp
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.cpp:805]: Why is this needed? Wasn't it recently deleted as unused?

**Comment by QuentinCaldeira-eaton** [comment]
It seems that remove ZephyrSocket.h, which implemented recvmsg function before zephyr implemented it, is still used in many examples. Should I put ZephyrSocket.h back ?

File: src/platform/Zephyr/CHIPDevicePlatformConfig.h
Comment (Damian-Nordic) [inline src/platform/Zephyr/CHIPDevicePlatformConfig.h:107]: `NEWLIB_LIBC_NANO` depends on `NEWLIB_LIBC` so this change is not needed, right?

File: src/inet/UDPEndPointImplSockets.cpp
Comment (Damian-Nordic) [inline src/inet/UDPEndPointImplSockets.cpp:784]: It's now discouraged to use `#ifdef` in the SDK - we build with `-Wundef` to catch typos in the preprocessor directives. Therefore, this should be added to `InetConfig.h` and defined as 0 by default (and properly documented). Then this could be overridden in the platform's configuration like you did.

File: src/platform/Zephyr/BLEManagerImpl.cpp
Comment (QuentinCaldeira-eaton) [reply src/platform/Zephyr/BLEManagerImpl.cpp:805]: I thought it was unused but see it implemented somewhere else, can remember. I can remove it

File: src/platform/Zephyr/CHIPDevicePlatformConfig.h
Comment (QuentinCaldeira-eaton) [reply src/platform/Zephyr/CHIPDevicePlatformConfig.h:107]: I guess, too, just need some feedbacks. It's like one of my first PR as an intern, and I probably will do some errors

File: src/inet/UDPEndPointImplSockets.cpp
Comment (QuentinCaldeira-eaton) [reply src/inet/UDPEndPointImplSockets.cpp:787]: Because I don't know how to get it. Do you have a method ?

File: src/inet/UDPEndPointImplSockets.cpp
Comment (QuentinCaldeira-eaton) [reply src/inet/UDPEndPointImplSockets.cpp:784]: Thanks, I didn't know if it was related to inet. How should I name it also ?

File: src/platform/Zephyr/CHIPDevicePlatformConfig.h
Comment (Damian-Nordic) [reply src/platform/Zephyr/CHIPDevicePlatformConfig.h:107]: That's OK, I was just curious if you had any build issues without it.

File: src/inet/UDPEndPointImplSockets.cpp
Comment (Damian-Nordic) [reply src/inet/UDPEndPointImplSockets.cpp:784]: You can follow the existing convention, something like `INET_CONFIG_UDP_SOCKET_MREQN` would probably work.

File: src/platform/Zephyr/DiagnosticDataProviderImpl.cpp
Comment (Damian-Nordic) [reply src/platform/Zephyr/DiagnosticDataProviderImpl.cpp:113]: Yes.

File: src/inet/UDPEndPointImplSockets.cpp
Comment (Damian-Nordic) [reply src/inet/UDPEndPointImplSockets.cpp:787]: aInterfaceId.GetPlatformInterface() perhaps?

**Comment by Damian-Nordic** [comment]
> It seems that remove ZephyrSocket.h, which implemented recvmsg function before zephyr implemented it, is still used in many examples. Should I put ZephyrSocket.h back ?

Oh yes, please. Just set CHIP_SYSTEM_CONFIG_USE_ZEPHYR_SOCKET_EXTENSIONS to 0 for Zephyr platform if needed.

**Comment by QuentinCaldeira-eaton** [comment]
@Damian-Nordic is there a clean way to put Zephyr socket back (like undo the [f283940](https://github.com/project-chip/connectedhomeip/pull/34135/commits/f283940c0312c9d4820a9e68610f319f356e45c8) commit) ?

File: src/platform/Zephyr/BLEManagerImpl.h
Comment (Damian-Nordic) [inline src/platform/Zephyr/BLEManagerImpl.h:74]: Please remove as well.

**Review by Damian-Nordic — APPROVED** [review]

**Comment by biii-anuj-pathak** [comment]
Hi @QuentinCaldeira-eaton 

First of all Thanks for filing this PR, I want to try Matter on ESP using Zephyr, and Luckily this PR came in right at the time for me.
I have followed the steps mentioned here https://github.com/project-chip/connectedhomeip/issues/33471. But i was not able to compile it for ESP.

Following are the steps
- included your fork in yaml file as per https://github.com/project-chip/connectedhomeip/issues/33471
- I have moved `config/zephyr/chip-module` 
- Changed paths in Kconfig, and CMakeLists.
- added CONFIG_CHIP=y

Following is the error I'm getting. 
```
[325/342] Performing build step for 'chip-gn'Starting Matter library build in D:/ZephyrApps/bfw/build/modules/connectedhomeip
diff: args.gn: No such file or directory
ERROR at //third_party/pigweed/repo/pw_toolchain/clang_tools.gni:15:1: Unable to load "D:/ZephyrApps/modules/lib/matter/build_overrides/pigweed_environment.gni".
import("//build_overrides/pigweed_environment.gni")
^-------------------------------------------------
See //third_party/pigweed/repo/pw_toolchain/host_clang/toolchains.gni:18:1: whence it was imported.
import("$dir_pw_toolchain/clang_tools.gni")
^-----------------------------------------
See //third_party/pigweed/repo/pw_unit_test/test.gni:22:1: whence it was imported.
import("$dir_pw_toolchain/host_clang/toolchains.gni")
^---------------------------------------------------
See //third_party/pigweed/repo/pw_string/BUILD.gn:21:1: whence it was imported.
import("$dir_pw_unit_test/test.gni")
^----------------------------------
See //src/lib/core/BUILD.gn:118:5: which caused the file to be included.
    "$dir_pw_string",
```

Let me know if i missed any step, or you could help me here. It will be really helpful

**Comment by QuentinCaldeira-eaton** [comment]
Hey @biii-anuj-pathak, I don't if it's the right place to discuss this, but here are the first things I think of.

- Did you clone the submodules as well (like pigweed, jsoncpp, etc...). You can do this with `git submodule update --recursive`
- Did you bootstrap after git clone ? It's mandatory
- Did you `source scripts/activate.sh` before building ?

By the way, if I can see the example you are building, it will help me to test this PR.
Tell me if it helps you !

File: src/platform/Zephyr/BLEManagerImpl.h
Comment (QuentinCaldeira-eaton) [reply src/platform/Zephyr/BLEManagerImpl.h]: Done, sorry I wasn't in office Friday

**Comment by QuentinCaldeira-eaton** [comment]
Hi @Damian-Nordic, I don't understand why nrf build doesn't pass. Any idea ?

**Comment by Damian-Nordic** [comment]
@QuentinCaldeira-eaton Most likely because you added `#include <zephyr/net/net_if.h>` to the common code without checking if Zephyr net_if usage is enabled in Matter (which in fact is not enabled for native_posix tests).

You can guard the include with `#if CHIP_DEVICE_CONFIG_ENABLE_WIFI` and it should help.

**Comment by QuentinCaldeira-eaton** [comment]
> @QuentinCaldeira-eaton Most likely because you added `#include <zephyr/net/net_if.h>` to the common code without checking if Zephyr net_if usage is enabled in Matter (which in fact is not enabled for native_posix tests).
> 
> You can guard the include with `#if CHIP_DEVICE_CONFIG_ENABLE_WIFI` and it should help.

Seems to the trick, thanks ! Now Darwin platform causing trouble. I don't if it's alarming or not.

**Comment by QuentinCaldeira-eaton** [comment]
@Damian-Nordic, I don't understand why Darwin failed. Any idea ?

**Comment by axelnxp** [comment]
Hi,

Could you give us (NXP) some time to review this PR ? We are also using zephyr with matter but we don't have workflows setup yet (coming soon, we are working on it). So we just need to check if it's ok on our end.

Thanks !

**Comment by QuentinCaldeira-eaton** [comment]
@axelnxp Yes, of course. Do I need to do something else or everything is okay ?

**Comment by axelnxp** [comment]
@QuentinCaldeira-eaton looks good to us, you can go ahead, thanks !

**Comment by QuentinCaldeira-eaton** [comment]
@axelnxp Thanks ! I'm just waiting for approval. If one day you have some time, could I discuss with you about Matter with Zephyr ?

**Review by andy31415 — APPROVED** [review]

**Pull request merged**
